multicluster/controllers/multicluster/member/serviceexport_controller.go

```diff
@@ -491,29 +503,67 @@ func (r *ServiceExportReconciler) updateSvcExportStatus(ctx context.Context, req
 func (r *ServiceExportReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	// Watch events only when resource version changes
-	versionChange := predicate.ResourceVersionChangedPredicate{}
+	versionChangePredicates := builder.WithPredicates(predicate.ResourceVersionChangedPredicate{})
 	if r.endpointSliceEnabled {
 		return ctrl.NewControllerManagedBy(mgr).
-			For(&k8smcsv1alpha1.ServiceExport{}).
-			Watches(&source.Kind{Type: &corev1.Service{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc)).
-			Watches(&source.Kind{Type: &discovery.EndpointSlice{}}, handler.EnqueueRequestsFromMapFunc(endpointSliceMapFunc)).
-			WithEventFilter(versionChange).
+			For(&k8smcsv1alpha1.ServiceExport{}, versionChangePredicates).
+			Watches(&source.Kind{Type: &corev1.Service{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc), versionChangePredicates).
+			Watches(&source.Kind{Type: &discovery.EndpointSlice{}}, handler.EnqueueRequestsFromMapFunc(endpointSliceMapFunc), versionChangePredicates).
+			Watches(&source.Kind{Type: &mcv1alpha2.ClusterSet{}}, handler.EnqueueRequestsFromMapFunc(r.clusterSetMapFunc),
+				builder.WithPredicates(statusReadyPredicate)).
 			WithOptions(controller.Options{
 				MaxConcurrentReconciles: common.DefaultWorkerCount,
 			}).
 			Complete(r)
 	}
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&k8smcsv1alpha1.ServiceExport{}).
-		Watches(&source.Kind{Type: &corev1.Service{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc)).
-		Watches(&source.Kind{Type: &corev1.Endpoints{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc)).
-		WithEventFilter(versionChange).
+		For(&k8smcsv1alpha1.ServiceExport{}, versionChangePredicates).
+		Watches(&source.Kind{Type: &corev1.Service{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc), versionChangePredicates).
+		Watches(&source.Kind{Type: &corev1.Endpoints{}}, handler.EnqueueRequestsFromMapFunc(objectMapFunc), versionChangePredicates).
+		Watches(&source.Kind{Type: &mcv1alpha2.ClusterSet{}}, handler.EnqueueRequestsFromMapFunc(r.clusterSetMapFunc),
+			builder.WithPredicates(statusReadyPredicate)).
 		WithOptions(controller.Options{
 			MaxConcurrentReconciles: common.DefaultWorkerCount,
 		}).
 		Complete(r)
 }
 
+// clusterSetMapFunc handles ClusterSet events by enqueuing all ServiceExports
+// into the reconciler processing queue.
+func (r *ServiceExportReconciler) clusterSetMapFunc(a client.Object) []reconcile.Request {
+	clusterSet := &mcv1alpha2.ClusterSet{}
+	requests := []reconcile.Request{}
+	if a.GetNamespace() != r.namespace {
+		return requests
+	}
+
+	ctx := context.TODO()
+	err := r.Client.Get(ctx, types.NamespacedName{Namespace: a.GetNamespace(), Name: a.GetName()}, clusterSet)
+	if err == nil {
+		if len(clusterSet.Status.Conditions) > 0 && clusterSet.Status.Conditions[0].Status == corev1.ConditionTrue {
+			svcExports := &k8smcsv1alpha1.ServiceExportList{}
+			r.Client.List(ctx, svcExports)
+			existingSvcExports := sets.Set[string]{}
```

Comment:
didn't get the purpose of `existingSvcExports`

Reply:
Stale variable, removed